alg/teca_2d_component_area.cxx

```diff
@@ -0,0 +1,352 @@
+#include "teca_2d_component_area.h"
+
+#include "teca_mesh.h"
+#include "teca_array_collection.h"
+#include "teca_variant_array.h"
+#include "teca_metadata.h"
+#include "teca_cartesian_mesh.h"
+
+#include <algorithm>
+#include <iostream>
+#include <deque>
+#include <set>
+#define _USE_MATH_DEFINES
+#include <cmath>
+
+#if defined(TECA_HAS_BOOST)
+#include <boost/program_options.hpp>
+#endif
+
+namespace {
+
+template <typename label_t>
+label_t get_max_label_id(unsigned long n, const label_t *labels)
+{
+    label_t max_label_id = std::numeric_limits<label_t>::lowest();
+    for (unsigned long i = 0; i < n; ++i)
+    {
+        label_t label = labels[i];
+        max_label_id = label > max_label_id ? label : max_label_id;
+    }
+    return max_label_id;
+}
+
+// visit each node in the mesh, the node is treated as a cell in
+// the dual mesh defined by mid points between nodes. the area of
+// the cell is added to the corresponding label. this formulation
+// requires a layer of ghost nodes.
+//
+// A = rho^2(cos(phi_0) - cos(phi_1))(theta_1 - theta_0)
+//
+template<typename coord_t, typename label_t, typename container_t>
+void component_area(unsigned long nx, unsigned long ny,
+    const coord_t *x, const coord_t *y, const label_t *labels,
+    container_t &area)
+{
+    // This calculation is sensative to floating point precision and
+    // should be done in double precision
+    using calc_t = double;
+
+    calc_t R_e = 6378.1370; // km
+    calc_t R_e_sq = R_e*R_e;
+    calc_t rad_per_deg = M_PI/180.0;
+
+    unsigned long nxm1 = nx - 1;
+    unsigned long nym1 = ny - 1;
+
+    // convert to spherical coordinates in units of radians,
+    // move to the dual mesh, and pre-compute factors in A
+    calc_t *rho_sq_d_theta = (calc_t*)malloc(nx*sizeof(calc_t));
+    rho_sq_d_theta[0] = calc_t();
+    for (unsigned long i = 1; i < nxm1; ++i)
+        rho_sq_d_theta[i] = R_e_sq*calc_t(0.5)*(x[i + 1] - x[i - 1])*rad_per_deg;
+    rho_sq_d_theta[nxm1] = calc_t();
+
+    calc_t *phi = (calc_t*)malloc(ny*sizeof(calc_t));
+    for (unsigned long j = 0; j < ny; ++j)
+        phi[j] = (calc_t(90.0) - y[j])*rad_per_deg;
+
+    calc_t *d_cos_phi = (calc_t*)malloc(ny*sizeof(calc_t));
+    for (unsigned long j = 1; j < nym1; ++j)
+    {
+        calc_t cos_phi_1 = cos(calc_t(0.5)*(phi[j - 1] + phi[j]));
+        calc_t cos_phi_0 = cos(calc_t(0.5)*(phi[j] + phi[j + 1]));
+        d_cos_phi[j] = cos_phi_0 - cos_phi_1;
+    }
+    d_cos_phi[0] = calc_t();
+    d_cos_phi[nym1] = calc_t();
+
+    // finish off the calc by multiplying the factors
+    for (unsigned long j = 1; j < nym1; ++j)
+    {
+        calc_t d_cos_phi_j = d_cos_phi[j];
+        unsigned long jj = j*nx;
+        for (unsigned long i = 1; i < nxm1; ++i)
+        {
+            area[labels[jj + i]] += rho_sq_d_theta[i]*d_cos_phi_j;
```

Comment:
This appears to be mathematically correct to me, but it is different from way I usually see the calculation formulated.  Did you use this approach for numerical stability reasons? 

Typically, for atmosphere/climate science applications, I see the area calculation written, equivalently, as `R**2 * d_lon_radians * d_lat_radians * cos(lat_radians)`.  That version of the formulation would reduce the number of calculations relative to what's here (only one cosine calculated, and no need for the pi-minus-lat calculation), but I'm not sure whether that approach might be numerically inferior for some reason.

Reply:
Haha, thanks, but in fact I used this formulation out of ignorance of what's typically done in the climate community. What I have is the correct way to formulate area on spherical patches, which can be derived directly using calculus. The formulation you mention is an approximation which leverages the fact that in this case because the radius is relatively large, and the patch is relatively small, the curvature could reasonably be neglected. One thing I like about the way I did it is that it's easy to test, as it's easy to compute the answer over a patch of any size, which gives you something to compare to. that's what I did in the regression test.  I don't mind formulating in the way you suggest if you think it would be better, or even having both formulations.